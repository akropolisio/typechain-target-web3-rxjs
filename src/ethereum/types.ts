import {
  EventLog as Web3EventLog,
  TransactionReceipt,
  PromiEvent,
  TransactionConfig,
} from 'web3-core';
import { O } from 'ts-toolbelt';
import { Observable } from 'rxjs';
import { Contract } from 'web3-eth-contract';
import BN from 'bn.js';

export interface ContractWrapper extends Omit<Contract, 'methods' | 'events' | 'getPastEvents'> {
  methods: Record<string, CallMethod<any, any> | SendMethod<any, any>>;
  events: Record<string, EventMethod<any, any>> & {
    allEvents: EventMethod<any, any>;
  };
  getPastEvents: PastEventsMethod<any, any>;
}

export type CallMethod<Input extends Arguments | void, Output extends void | ArgumentType> = (
  input: Input,
  eventsForReload?: EventEmitter<any> | EventEmitter<any>[],
  tx?: CallOptions,
) => Observable<Output>;

export type SendMethod<Input extends Arguments | void, Output extends void | ArgumentType> = ((
  input: Input,
  tx: SendOptions,
) => PromiEvent<TransactionReceipt>) & {
  getTransaction(input: Input): TransactionObject;
  read: ReadMethod<Input, Output>;
};

export type ReadMethod<Input extends Arguments | void, Output extends void | ArgumentType> = (
  input: Input,
  tx: SendOptions,
  eventsForReload?: EventEmitter<any> | EventEmitter<any>[],
) => Observable<Output>;

type EventMethod<Input extends Arguments, IndexedInput extends Arguments> = (
  options?: SubscribeEventOptions<IndexedInput>,
  cb?: Callback<EventLog<Input>>,
) => EventEmitter<Input>;

type PastEventsMethod<
  Inputs extends Record<string, Arguments>,
  IndexedInputs extends Record<keyof Inputs, Arguments>
> = {
  <EventName extends keyof Inputs>(event: EventName): Promise<EventLog<Inputs[EventName]>[]>;
  <EventName extends keyof Inputs>(
    event: EventName,
    options: PastEventsOptions<IndexedInputs[EventName]>,
    callback: (error: Error, event: EventLog<Inputs[EventName]>) => void,
  ): Promise<EventLog<Inputs[EventName]>[]>;
  <EventName extends keyof Inputs>(
    event: EventName,
    options: PastEventsOptions<IndexedInputs[EventName]>,
  ): Promise<EventLog<Inputs[EventName]>[]>;
  <EventName extends keyof Inputs>(
    event: EventName,
    callback: (error: Error, event: EventLog<Inputs[EventName]>) => void,
  ): Promise<EventLog<Inputs[EventName]>[]>;
};

type PastEventsOptions<IndexedInput extends Arguments> = SubscribeEventOptions<IndexedInput> & {
  toBlock?: BlockType;
};

interface SubscribeEventOptions<IndexedInput extends Arguments> {
  filter?: Partial<IndexedInput>;
  fromBlock?: BlockType;
  topics?: string[];
}

export type Arguments = Record<string, ArgumentType>;
export type ArgumentType = PlainArgumentType | PlainArgumentType[];
export type PlainArgumentType = JSType | { [key in string]: PlainArgumentType };

type BlockType = 'latest' | 'pending' | 'genesis' | number;
type Callback<T> = (error: Error, result: T) => void;

/* ** EXAMPLE START ** */

interface X extends ContractWrapper {
  methods: {
    testCall: CallMethod<void, void>;
    testCall1: CallMethod<{ amount: BN }, string>;
    testCall2: CallMethod<void, [string, BN]>;
    testSend: SendMethod<void, void>;
    testSend1: SendMethod<{ amount: BN }, string>;
    testSend2: SendMethod<void, [string, BN]>;
  };
  events: {
    TestEvent: EventMethod<EventInput, EventIndexedInput>;
    TestEvent1: EventMethod<Event1Input, Event1IndexedInput>;
    TestEvent2: EventMethod<Event2Input, Event2IndexedInput>;
    allEvents: EventMethod<
      EventInput | Event1Input | Event2Input,
      EventIndexedInput | EventIndexedInput | EventIndexedInput
    >;
  };
  getPastEvents: PastEventsMethod<
    {
      TestEvent: EventInput;
      TestEvent1: Event1Input;
      TestEvent2: Event2Input;
    },
    {
      TestEvent: EventIndexedInput;
      TestEvent1: Event1IndexedInput;
      TestEvent2: Event2IndexedInput;
    }
  >;
}

type EventInput = {
  amount: BN;
  user: string;
};
type EventIndexedInput = {
  user: string;
};
type Event1Input = {
  total: BN;
  user: string;
};
type Event1IndexedInput = {
  user: string;
};
type Event2Input = {
  user: string;
  amount: BN;
};
type Event2IndexedInput = {
  amount: BN;
};

const x: X = null as any;

const a = x.methods.testCall();
const b = x.methods.testCall1({ amount: new BN(0) });
const c = x.methods.testCall2(undefined, x.events.TestEvent({ filter: { user: '0x00' } }));

const a1 = x.methods.testSend(undefined, { from: '0x00' });
const b1 = x.methods.testSend1({ amount: new BN(0) }, { from: '0x00' });
const c1 = x.methods.testSend2(undefined, { from: '0x00' });

const a2 = x.methods.testSend.getTransaction();
const b2 = x.methods.testSend1.getTransaction({ amount: new BN(0) });
const c2 = x.methods.testSend2.getTransaction();

const a3 = x.methods.testSend.read(undefined, { from: '0x00' });
const b3 = x.methods.testSend1.read({ amount: new BN(0) }, { from: '0x00' });
const c3 = x.methods.testSend2.read(
  undefined,
  { from: '0x00' },
  x.events.TestEvent({ filter: { user: '0x00' } }),
);

const d = x.events.TestEvent2({ filter: { amount: new BN(0) } });
const e = x.events.allEvents({ filter: { user: '0x00' } });

const f = x.getPastEvents('TestEvent', {
  filter: { user: '0x00' },
  fromBlock: 123,
  toBlock: 'latest',
});

/* ** EXAMPLE END ** */

export interface Web3ContractInputArgs extends Array<JSType | Web3ContractInputArgs> {}
export type Web3ContractOutputArgs = JSType | Web3ContractInputArgs;

export interface TransactionObject {
  arguments: any[];
  call(tx?: CallOptions): Promise<TransactionReceipt>;
  send: {
    (tx?: SendOptions): PromiEvent<TransactionReceipt>;
    request(
      options: Omit<SendOptions, 'to' | 'data'>,
    ): { params: [TransactionConfig]; method: string };
  };
  estimateGas(tx?: SendOptions): Promise<number>;
  encodeABI(): string;
}

export type EventLog<T> = Omit<Web3EventLog, 'returnValues'> & { returnValues: T };

export interface EventEmitter<T> {
  on(type: 'data', handler: (event: EventLog<T>) => void): EventEmitter<T>;
  on(type: 'changed', handler: (receipt: EventLog<T>) => void): EventEmitter<T>;
  on(type: 'error', handler: (error: Error) => void): EventEmitter<T>;
  on(
    type: 'error' | 'data' | 'changed',
    handler: (error: Error | TransactionReceipt | string) => void,
  ): EventEmitter<T>;
}

export type SendOptions = O.Required<CallOptions, 'from'>;

export interface CallOptions {
  nonce?: string | number;
  chainId?: string | number;
  from?: string;
  value?: string | number;
  gas?: string | number;
  gasPrice?: string | number;
  maxPriorityFeePerGas?: string | number;
  maxFeePerGas?: string | number;
}

export type InputEvmTypeToJSTypeMap = {
  address: string;
  integer: BN;
  uinteger: BN;
  boolean: boolean;
  string: string;
  bytes: string;
  'dynamic-bytes': string;
};

export type InputEvmType = keyof InputEvmTypeToJSTypeMap;

export type JSType = InputEvmTypeToJSTypeMap[InputEvmType];
