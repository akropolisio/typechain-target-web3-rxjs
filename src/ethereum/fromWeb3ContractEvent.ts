import { fromEventPattern, Observable } from 'rxjs';
import { EventData } from 'web3-eth-contract';

type EventEmitter = {
  on(event: string, handler: (event: EventData) => void): EventEmitter;
};

export function fromWeb3DataEvent(emitter: EventEmitter): Observable<EventData> {
  interface IUnsubscribable {
    unsubscribe: () => void;
  }

  return fromEventPattern<EventData>(
    handler => emitter.on('data', handler),
    (_, signal: IUnsubscribable) => signal.unsubscribe(),
  );
}
