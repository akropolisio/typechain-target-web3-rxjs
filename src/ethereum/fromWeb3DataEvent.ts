import { fromEventPattern, Observable } from 'rxjs';
import { EventLog } from 'web3-core';

import { EventEmitter } from './types';

export function fromWeb3DataEvent(emitter: EventEmitter<any>): Observable<EventLog> {
  interface IUnsubscribable {
    unsubscribe: () => void;
  }

  return fromEventPattern<EventLog>(
    handler => emitter.on('data', handler as (event: EventLog) => void),
    (_, signal: IUnsubscribable) => signal.unsubscribe(),
  );
}
