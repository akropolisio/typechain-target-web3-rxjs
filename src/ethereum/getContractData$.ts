import { Observable, from, merge, EMPTY, ReplaySubject } from 'rxjs';
import { skipUntil, mergeMap, throttleTime, switchMap, shareReplay } from 'rxjs/operators';
import { Eth } from 'web3-eth';
import { Contract } from 'web3-eth-contract';

import {
  EventLog,
  CallOptions,
  Web3ContractInputArgs,
  Web3ContractResponse,
  Response,
} from './types';

interface IOptions {
  eventsForReload?: Observable<EventLog<any>> | Observable<EventLog<any>>[];
  reloadTrigger$?: Observable<any>;
  args?: Web3ContractInputArgs;
  convert?(value: Web3ContractResponse): Response | void;
  tx?: CallOptions;
}

function identity(value: any) {
  return value;
}

function awaitMs(delayMs: number): Promise<void> {
  return new Promise(resolve => setTimeout(resolve, delayMs));
}

export function getContractData$(
  contract: Contract,
  eth: Eth,
  method: string,
  options: IOptions = {},
): Observable<Response> {
  const { eventsForReload = [], reloadTrigger$ = EMPTY, args = [], convert = identity } = options;

  const load = async () => {
    const data = await contract.methods[method](...args).call(options.tx);
    return convert(data);
  };

  const reloadEvents = Array.isArray(eventsForReload) ? eventsForReload : [eventsForReload];

  const first$ = from(load());
  const fromEvents$ = merge(...reloadEvents).pipe(
    skipUntil(first$),
    throttleTime(200),
    switchMap(async event => {
      let currentBlock = await eth.getBlockNumber();

      /* eslint-disable no-await-in-loop */
      while (currentBlock < event.blockNumber) {
        await awaitMs(500);
        currentBlock = await eth.getBlockNumber();
      }
      /* eslint-enable no-await-in-loop */

      return event;
    }),
    mergeMap(() => from(load()), 1),
    shareReplay(1),
  );

  const subject = new ReplaySubject<Response>(1);

  merge(first$, fromEvents$, reloadTrigger$).subscribe(subject);

  return subject;
}
