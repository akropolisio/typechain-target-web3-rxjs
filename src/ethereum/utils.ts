import { RawAbiDefinition, RawAbiParameter } from 'typechain';

export function getWeb3Signature(abi: Pick<RawAbiDefinition, 'name' | 'inputs'>): string {
  return `${abi.name}(${abi.inputs.map(i => getArgumentForSignature(i)).join(',')})`;
}

function getArgumentForSignature(argument: RawAbiParameter): string {
  if (argument.type === 'tuple') {
    return `(${argument.components?.map(i => getArgumentForSignature(i)).join(',')})`;
  }
  if (argument.type === 'tuple[]') {
    return `${getArgumentForSignature({ ...argument, type: 'tuple' })}[]`;
  }
  return argument.type;
}
