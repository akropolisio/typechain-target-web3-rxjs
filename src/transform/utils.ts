import { O } from 'ts-toolbelt';
import {
  AbiOutputParameter,
  AbiParameter,
  EventDeclaration,
  EvmOutputType,
  FunctionDeclaration,
  getSignatureForFn,
  TupleType,
} from 'typechain';

export function unfoldDeclarations<T extends FunctionDeclaration | EventDeclaration>(
  declaration: Record<string, T[]>,
): T[] {
  const toOverloadedDeclaration = (entry: T) => ({
    ...entry,
    name: `"${getSignatureForFn(entry as FunctionDeclaration)}"`,
  });

  return Object.values(declaration).flatMap(entries =>
    entries.length === 1 ? entries[0] : [entries[0], ...entries.map(toOverloadedDeclaration)],
  );
}

export function codegenInputTypes(inputs: O.Optional<AbiParameter, 'name'>[]): string {
  if (inputs.length === 0) {
    return 'void';
  }
  return `{
    ${inputs
      .map((input, index) => `${input.name || `arg${index}`}: ${codegenType(input.type)}`)
      .join(';\n')}
  }`;
}

export function codegenOutputTypes(outputs: O.Optional<AbiOutputParameter, 'name'>[]): string {
  if (outputs.length === 1) {
    return codegenType(outputs[0].type);
  }
  return `{
      ${outputs.map(t => t.name && `${t.name}: ${codegenType(t.type)};\n`).join('')}
      ${outputs.map((t, i) => `${i}: ${codegenType(t.type)}`).join(';\n')}
    }`;
}

export function codegenType(evmType: EvmOutputType): string {
  switch (evmType.type) {
    case 'integer':
    case 'uinteger':
      return 'BN';
    case 'address':
      return 'string';
    case 'bytes':
    case 'dynamic-bytes':
      return 'string';
    case 'array':
      return `(${codegenType(evmType.itemType)})[]`;
    case 'boolean':
      return 'boolean';
    case 'string':
      return 'string';
    case 'tuple':
      return codegenTupleType(evmType);
    case 'void':
      return 'void';
    case 'unknown':
    default:
      throw new Error(`Unhandled type while code generation: ${evmType.originalType}`);
  }
}

export function codegenTupleType(tuple: TupleType) {
  return `{
    ${tuple.components
      .map(component => `${component.name}: ${codegenType(component.type)}`)
      .join(';\n')}
  }`;
}

export function uniq<T>(array: T[]): T[] {
  return array.reduce((acc, curr) => (acc.includes(curr) ? acc : [...acc, curr]), [] as T[]);
}
