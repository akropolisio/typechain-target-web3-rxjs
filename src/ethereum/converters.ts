import BN from 'bn.js';
import { RawAbiParameter } from 'typechain';

import {
  JSType,
  EvmType,
  Arguments,
  Response,
  InputEvmTypeToJSTypeMap,
  OutputEvmTypeToJSTypeMap,
  Web3ContractInputArgs,
  Web3ContractResponse,
  ArgumentType,
  isPlainValue,
} from './types';
import { attachStaticFields } from './utils';

export function inputsToRequest(
  inputsAbi: RawAbiParameter[],
  args: Arguments,
): Web3ContractInputArgs {
  return inputsAbi.map((abiParameter, index) => {
    const { isArray, type } = parseEvmType(abiParameter.type);
    const arg = (abiParameter.name && args[abiParameter.name]) || args[index];
    if (!arg) {
      throw new Error(`ABI parameter ${abiParameter.name} not found in ${args}`);
    }

    if (isArray) {
      if (!Array.isArray(arg)) {
        throw new Error(`Expected an array, but received ${arg}`);
      }
      return arg.map(innerArgs =>
        type === 'tuple'
          ? convertTupleValueToRequest(abiParameter, innerArgs)
          : convertPlainValueToRequest(type, innerArgs),
      );
    }

    return type === 'tuple'
      ? convertTupleValueToRequest(abiParameter, arg)
      : convertPlainValueToRequest(type, arg);
  });
}

function convertTupleValueToRequest(abiParameter: RawAbiParameter, args: ArgumentType) {
  if (!abiParameter.components || isPlainValue(args)) {
    throw new Error(`Unable to convert ${args} to a tuple request`);
  }
  return inputsToRequest(abiParameter.components, args as Arguments);
}

function convertPlainValueToRequest(type: EvmType, inputValue: ArgumentType): JSType {
  if (!isPlainValue(inputValue)) {
    throw new Error(`Expected string, boolean, or BN, but received ${inputValue}`);
  }

  const toRequest: {
    [key in EvmType]: (input: InputEvmTypeToJSTypeMap[key]) => JSType;
  } = {
    address: value => value,
    boolean: value => value,
    integer: value => value.toString(),
    uinteger: value => value.toString(),
    string: value => value,
    bytes: value => value,
    'dynamic-bytes': value => value,
  };

  return (toRequest[type] as any)(inputValue);
}

export function responseToOutput(
  outputsAbi: RawAbiParameter[],
  response: Web3ContractResponse,
): Response {
  if (outputsAbi.length === 1) {
    return convertResponseValueToOutput(outputsAbi[0], response);
  }

  return toMixedArray(outputsAbi, response);
}

function convertResponseValueToOutput(
  abiParameter: RawAbiParameter,
  responseValue: Web3ContractResponse,
): Response {
  const { type, isArray } = parseEvmType(abiParameter.type);

  if (isArray) {
    if (!Array.isArray(responseValue)) {
      throw new Error(`Expected an array, but received ${responseValue}`);
    }

    return responseValue.map(innerValues =>
      type === 'tuple'
        ? convertTupleResponseToOutput(abiParameter, innerValues)
        : convertPlainResponseToOutput(type, innerValues),
    );
  }
  return type === 'tuple'
    ? convertTupleResponseToOutput(abiParameter, responseValue)
    : convertPlainResponseToOutput(type, responseValue);
}

function convertTupleResponseToOutput(
  abiParameter: RawAbiParameter,
  responseValue: Web3ContractResponse,
) {
  if (!abiParameter.components) {
    throw new Error(`Components property is required for tuples`);
  }
  return toMixedArray(abiParameter.components, responseValue);
}

function toMixedArray(abiParameters: RawAbiParameter[], value: Web3ContractResponse) {
  if (isPlainValue(value)) {
    throw new Error(`Expected an object, or an array, but received ${value}`);
  }
  return abiParameters.reduce((resultAcc, componentAbi, index) => {
    const convertedValue = convertResponseValueToOutput(componentAbi, value[index]);
    resultAcc.push(convertedValue);
    if (componentAbi.name) {
      attachStaticFields(resultAcc, { [componentAbi.name]: convertedValue });
    }
    return resultAcc;
  }, [] as Response[]);
}

function convertPlainResponseToOutput(type: EvmType, value: Web3ContractResponse): JSType {
  if (!isPlainValue(value)) {
    throw new Error(`Expected string, boolean, or BN, but received ${value}`);
  }

  const toOutput: {
    [key in EvmType]: (typedValue: OutputEvmTypeToJSTypeMap[key]) => JSType;
  } = {
    address: v => v,
    boolean: v => v,
    integer: v => new BN(v),
    uinteger: v => new BN(v),
    string: v => v,
    bytes: v => v,
    'dynamic-bytes': v => v,
  };

  return (toOutput[type] as any)(value);
}

function parseEvmType(abiType: string): { type: EvmType | 'tuple'; isArray: boolean } {
  const [, inputType, isArray] = abiType.match(/^(.+?)(\[\d*?\])?$/) || [];

  const type = ((): EvmType | 'tuple' => {
    // eslint-disable-next-line default-case
    switch (inputType) {
      case 'bool':
        return 'boolean';
      case 'address':
        return 'address';
      case 'string':
        return 'string';
      case 'byte':
        return 'bytes';
      case 'bytes':
        return 'dynamic-bytes';
    }

    if (inputType.startsWith('tuple')) {
      return 'tuple';
    }
    if (inputType.startsWith('uint')) {
      return 'uinteger';
    }
    if (inputType.startsWith('int')) {
      return 'integer';
    }
    if (inputType.startsWith('bytes')) {
      return 'bytes';
    }

    throw new Error(`Type "${abiType}" cannot be parsed`);
  })();

  return { type, isArray: !!isArray };
}
