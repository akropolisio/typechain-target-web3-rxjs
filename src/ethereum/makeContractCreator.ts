import { Observable } from 'rxjs';
import Web3 from 'web3';
import { Contract } from 'web3-eth-contract';
import { RawAbiDefinition, RawAbiParameter } from 'typechain';

import { getContractData$ } from './getContractData$';
import {
  EventEmitter,
  SendOptions,
  CallOptions,
  InputEvmType,
  InputEvmTypeToJSTypeMap,
  JSType,
  ContractWrapper,
  Web3ContractInputArgs,
  CallMethod,
  ReadMethod,
  SendMethod,
  Arguments,
  PlainArgumentType,
} from './types';
import { getWeb3Signature } from './utils';

const toRequest: {
  [key in InputEvmType]: (input: InputEvmTypeToJSTypeMap[key]) => JSType;
} = {
  address: value => value,
  boolean: value => value,
  integer: value => value.toString(),
  uinteger: value => value.toString(),
  string: value => value,
  bytes: value => value,
  'dynamic-bytes': value => value,
};

export function makeContractCreator<C extends ContractWrapper>(abi: any[]) {
  return (web3: Web3, address: string): C => {
    const baseContract = new web3.eth.Contract(abi, address);

    const methodsProxy = new Proxy<Contract['methods']>(
      {},
      {
        get(target: Contract['methods'], prop: string) {
          const methodAbi = extractAbiMethod(abi, prop);

          if (!methodAbi) {
            return target[prop];
          }

          const baseCallFunction = (
            input: void | Arguments,
            eventsForReload?: EventEmitter<any> | EventEmitter<any>[],
            tx?: CallOptions,
          ) => {
            return getContractData$(baseContract, web3.eth, prop, {
              tx,
              args: input ? inputsToRequest(methodAbi.inputs, input) : [],
              eventsForReload,
              // convert: (result) => responseToOutput(methodAbi.outputs, result),
            }) as Observable<any>;
          };

          const callFunction: CallMethod<any, any> = baseCallFunction;

          const readFunction: ReadMethod<any, any> = (
            input: void | Arguments,
            tx: SendOptions,
            eventsForReload?: EventEmitter<any> | EventEmitter<any>[],
          ) => baseCallFunction(input, eventsForReload, tx);

          if (methodAbi.stateMutability === 'view') {
            return callFunction;
          }

          const getTransactionFunction = (input: void | Arguments) => {
            const args = input ? inputsToRequest(methodAbi.inputs, input) : [];
            return baseContract.methods[prop](...args);
          };

          const sendFunction: SendMethod<any, any> = attachStaticFields(
            (input: void | Arguments, tx: SendOptions) => {
              return getTransactionFunction(input).send(tx);
            },
            {
              getTransaction: getTransactionFunction,
              read: readFunction,
            },
          );

          return sendFunction;
        },
      },
    );

    return (new Proxy<Contract>(baseContract, {
      get(target, prop: keyof Contract) {
        if (prop === 'methods') {
          return methodsProxy;
        }
        return target[prop];
      },
    }) as unknown) as C;
  };
}

export function inputsToRequest(
  inputsAbi: RawAbiParameter[],
  args: Arguments,
): Web3ContractInputArgs {
  return inputsAbi.map(parameterAbi => {
    const { isArray, type } = parseEvmType(parameterAbi.type);

    if (type === 'tuple') {
      return isArray
        ? (args[parameterAbi.name] as PlainArgumentType[]).map(innerArgs =>
            inputsToRequest(parameterAbi.components || [], innerArgs as Arguments),
          )
        : inputsToRequest(parameterAbi.components || [], args[parameterAbi.name] as Arguments);
    }

    return isArray
      ? (args[parameterAbi.name] as PlainArgumentType[]).map(innerArgs =>
          convertInputValueToRequest(type, innerArgs as JSType),
        )
      : convertInputValueToRequest(type, args[parameterAbi.name] as JSType);
  });
}

function convertInputValueToRequest(type: InputEvmType, inputValue: JSType): JSType {
  return (toRequest[type] as any)(inputValue);
}

function parseEvmType(abiType: string): { type: InputEvmType | 'tuple'; isArray: boolean } {
  const [, inputType, isArray] = abiType.match(/^(.+?)(\[\])?$/) || [];

  const type = ((): InputEvmType | 'tuple' => {
    // eslint-disable-next-line default-case
    switch (abiType) {
      case 'bool':
        return 'boolean';
      case 'address':
        return 'address';
      case 'string':
        return 'string';
      case 'byte':
        return 'bytes';
      case 'bytes':
        return 'dynamic-bytes';
    }

    if (inputType.startsWith('tuple')) {
      return 'tuple';
    }
    if (inputType.startsWith('uint')) {
      return 'uinteger';
    }
    if (inputType.startsWith('int')) {
      return 'integer';
    }
    if (inputType.startsWith('bytes')) {
      return 'bytes';
    }

    throw new Error(`Type "${abiType}" cannot be parsed`);
  })();

  return { type, isArray: !!isArray };
}

export function attachStaticFields<T extends {}, I extends Record<string, any>>(
  target: T,
  staticFields: I,
): T & I {
  const result: T & I = target as T & I;

  Object.keys(staticFields).forEach((key: keyof I) => {
    (result as I)[key] = staticFields[key];
  });

  return result;
}

function extractAbiMethod(abi: RawAbiDefinition[], method: string): RawAbiDefinition | null {
  const [, methodName] = method.match(/(\w+?)($|\(.*)/) || [];

  if (!methodName) {
    console.warn(new Error(`Invalid method: ${method}`));
    return null;
  }

  const methodEntries = abi.filter(m => m.name === methodName && m.type === 'function');

  if (!methodEntries.length) {
    console.warn(new Error(`Method ${method} not found`));
    return null;
  }

  const isSimpleMethodName = method === methodName;
  if (isSimpleMethodName) {
    return methodEntries[0];
  }

  const entry = methodEntries.find(m => getWeb3Signature(m) === method);

  if (!entry) {
    console.warn(new Error(`Method ${method} not found`));
    return null;
  }

  return entry;
}
