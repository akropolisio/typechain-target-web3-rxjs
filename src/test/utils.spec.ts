import { getWeb3Signature } from 'ethereum/utils';

import testAbi from './test.json';

describe('getWeb3Signature', () => {
  it('should support tuple and tuple[] types', () => {
    expect(testAbi.map(abi => getWeb3Signature(abi as any))).toMatchSnapshot();
  });
});
