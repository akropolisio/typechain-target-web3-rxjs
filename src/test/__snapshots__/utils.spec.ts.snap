// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getWeb3Signature should support tuple and tuple[] types 1`] = `
Array [
  "testRead((bool,address[],(uint256,uint256)[]),(bytes,string),uint256)",
  "testRead()",
  "testSend((uint256,uint256[],(uint256,uint256)[]),(uint256,uint256),uint256)",
  "Event1(uint256,bytes32)",
  "Event1(uint256,(uint256,uint256)[])",
  "Event2(uint256,bytes32)",
]
`;
